### PR TITLE
Remove Restricted String Mapping Param

### PR DESCRIPTION
**Pull request by original-brownbear**
This param was incredibly expensive to set up when parsing mappings and
is one of the big contributors to mapping parsing slowness on master.
Since all uses of this parameter type are statically known it seems the most
straight forward to simply statically hard code the validators so that we save
some allocations.

relates https://github.com/elastic/elasticsearch/issues/77466

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/TextParams.java [hunk 1]
```diff
@@ -129,7 +129,33 @@ private NamedAnalyzer wrapAnalyzer(NamedAnalyzer a) {
     }
 
     public static Parameter<String> indexOptions(Function<FieldMapper, String> initializer) {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/TextParams.java:131]: I think I'd prefer these to be `keywordIndexOptions` and `textIndexOptions`, as elsewhere `indexOptions` implies the full set of lucene options.

**Review by romseygeek — APPROVED** [review]
LGTM. Interesting that this is such a performance hit, do you know what bit of it in particular was slowing things down?

File: server/src/main/java/org/elasticsearch/index/mapper/TextParams.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/mapper/TextParams.java]: ++ renamed :)

**Comment by original-brownbear** [comment]
Thanks Alan!

The problem here performance wise is during serialisation of mappings. We have a massive number of these parameters and they're not free to instantiate. The creation of a linked hash map + capturing lambda starts to really hurt if you do it millions of times potentially:

image

The other big mountain is restricted string param as well. That said, the params instantiation still isn't free and we can do even better here I think. I have some follow-up ideas but this one already saves me a couple of minutes per benchmark run potentially :)

**Pull request merged**